# AM10IM - Introduction to Matlab

## Lecture Challenge 2 - Solutions

The idea behind this lecture challenge is to get used to manipulating if statements, and we apply this to the 2008 Olympic medal table and possible point systems.

In [1]:
%Country = [G S B]
US = [36 39 37]; %USA medals
GB = [19 13 19]; %Great Britain
CH = [48 22 30]; %China
RU = [24 13 23]; %Russia

These were the final results of the 2008 Olympics, but who 'won' the Olympics?

The different point systems that are in use are:

In [3]:
%Weighting [G S B]
W1 = [3 2 1]; % Simple point system
W2 = [4 2 1]; % New York Times point system
W3 = [5 3 1]; % 1908 London point system

__Challenge 1__ Using these different weights, can you find the score for each country for each different point system?

In [9]:
US1 = US*W1'
GB1 = GB*W1'
CH1 = CH*W1'
RU1 = RU*W1'

US1 =  223
GB1 =  102
CH1 =  218
RU1 =  121


Provides us with the results using the Simple Point weighting system. The "*" performs matrix multiplication, and the top line was equivalent to:
\begin{align}
  \text{US1} =& \begin{matrix}\begin{pmatrix}36 & 39 & 37\end{pmatrix}\\\mbox{}\end{matrix} 
  \begin{pmatrix} 3 \\ 2 \\ 1 \end{pmatrix} \\
  =& 36\times3 + 39\times2 + 37\times1
\end{align}

Notice that we had to use the transpose function:

In [8]:
W1'

ans =

   3
   2
   1



So that matrix multiplication could be performed.

Similarly for the other weights:

In [10]:
US2 = US*W2';
GB2 = GB*W2';
RU2 = RU*W2';
CH2 = CH*W2';

In [11]:
US3 = US*W3';
GB3 = GB*W3';
RU3 = RU*W3';
CH3 = CH*W3';

To save us from writing 12 lines of very similar code, we could have performed something along the lines of:

In [12]:
W = [W1' W2' W3']

W =

   3   4   5
   2   2   3
   1   1   1



Which puts the weights together in a matrix (each column is a different point system). Then put the medals together in rows:

In [13]:
M = [US;GB;RU;CH]

M =

   36   39   37
   19   13   19
   24   13   23
   48   22   30



Then the points for each team in each system is:

In [15]:
R = M*W

R =

   223   259   334
   102   121   153
   121   145   182
   218   266   336



Where matrix multiplication has been performed:
\begin{align}
R =& \begin{pmatrix}
36 & 39 & 37 \\
19 & 13 & 19 \\
24 & 13 & 23 \\
48 & 22 & 30
\end{pmatrix}
\begin{pmatrix}
3 & 4 & 5 \\
2 & 2 & 3 \\
1 & 1 & 1
\end{pmatrix}\\
=& \begin{pmatrix}
223 & 259 & 334 \\
102 & 121 & 153 \\
121 & 145 & 182 \\
218 & 266 & 336
\end{pmatrix}
\end{align}

The first column of R is the US, GB, RU, CH points in the __simple__ point system, the second column of R is the US, GB, RU, CH points in the __New York Times__ point system and the last column of R is the US, GB, RU and CH points in the __1908 London__ point system. This was three lines of code instead of twelve.

__Challenge 2__ Using If statements, can you find which team came first and which team came last in each point system?

Using if statements this is a long amount of code as it requires us to think about nested if statements for each point system. If you can understand how this works, then you understand if statements: The code is the same for each point system just swap the labels for each country.

In [17]:
if (US1 > GB1) && (US1 > RU1) && (US1 > CH1)
    fprintf('US came first in the simple point system with %.0f points \n',US1)
elseif (GB1 > US1) && (GB1 > RU1) && (GB1 > CH1)
    fprintf('GB came first in the simple point system with %.0f points \n',GB1)
elseif (CH1 > US1) && (CH1 > RU1) && (CH1 > GB1)
    fprintf('CH came first in the simple point system with %.0f points \n',CH1)
else
    fprintf('RU came first in the simple point system with %.0f points \n',RU1)
end

US came first in the simple point system with 223 points 


In [18]:
if (US2 > GB2) && (US2 > RU2) && (US2 > CH2)
    fprintf('US came first in the New York Times system with %.0f points \n',US2)
elseif (GB2 > US2) && (GB2 > RU2) && (GB2 > CH2)
    fprintf('GB came first in the New York Times system with %.0f points \n',GB2)
elseif (CH2 > US2) && (CH2 > RU2) && (CH2 > GB2)
    fprintf('CH came first in the New York Times system with %.0f points \n',CH2)
else
    fprintf('RU came first in the New York Times system with %.0f points \n',RU2)
end

CH came first in the New York Times system with 266 points 


In [19]:
if (US3 > GB3) && (US3 > RU3) && (US3 > CH3)
    fprintf('US came first in the 1908 London point system with %.0f points \n',US2)
elseif (GB3 > US3) && (GB3 > RU3) && (GB3 > CH3)
    fprintf('GB came first in the 1908 London point system with %.0f points \n',GB2)
elseif (CH3 > US3) && (CH3 > RU3) && (CH3 > GB3)
    fprintf('CH came first in the 1908 London point system with %.0f points \n',CH2)
else
    fprintf('RU came first in the 1908 London point system with %.0f points \n',RU2)
end

CH came first in the 1908 London point system with 266 points 


Hopefully you can see that this is __horribly inefficient code__, if I wanted to introduce a new point system I'd have to write another identical if statement. However, our goal here was not to make efficient code, but to learn the structure and understand if statements. To get the last place we simply switch the '>' in the if statements to '<'

Now that we know how for loops work, we can use a more efficient bit of code:

In [52]:
for i = 1:3
    [x1 y1] = max(R(:,i)); %Here x1 is the maximum value and y1 is the position of the maximum.
    if y1 == 1
        fprintf('In System %.0f US came first with %.0f points \n',i, R(y1,i))
    elseif y1 == 2
        fprintf('In System %.0f GB came first with %.0f points \n',i, R(y1,i))
    elseif y1 == 3
        fprintf('In System %.0f RU came first with %.0f points \n',i, R(y1,i))
    else
        fprintf('In System %.0f CH came first with %.0f points \n',i, R(y1,i))
    end    
    [x2 y2] = min(R(:,1)); % Here x2 is the minimum value and y2 is the position of the minimum.
    if y2 == 1
        fprintf('In System %.0f US came last with %.0f points \n',i, R(y2,i))
    elseif y2 == 2
        fprintf('In System %.0f GB came last with %.0f points \n',i, R(y2,i))
    elseif y2 == 3
        fprintf('In System %.0f RU came last with %.0f points \n',i, R(y2,i))
    else
        fprintf('In System %.0f CH came last with %.0f points \n',i, R(y2,i))
    end
    fprintf('\n') % an extra return to split the results
end

In System 1 US came first with 223 points 
In System 1 GB came last with 102 points 

In System 2 CH came first with 266 points 
In System 2 GB came last with 121 points 

In System 3 CH came first with 336 points 
In System 3 GB came last with 153 points 



Did you notice that I put __two__ numbers into fprintf? The first percentage sign corresponded to the first number after the quotation marks and the second percentage sign corresponded to the second number after the quotation marks.

__Challenge 3__ Can you now __rank__ the teams in order of highest to lowest points in each point system?

Again this is to simply see the structure of if statements. The following code is __totally unmanageble__, don't ever write anything like this outside of this challenge, but can you follow the logic?

In [35]:
if (US1 > CH1) && (US1 > RU1)  && (US1>GB1)
    USR1 = 1 % US greater than CH, RU and GB
    if (CH1 > RU1) && (CH1 > GB1)
        CHR1 = 2 %CH less than US but greater than RU and GB
        if (RU1 > GB1) %RU greater than GB but less than US and CH
            RUR1 = 3
            GBR1 = 4
        else           %GB greater than RU but less than US and CH 
            GBR1 = 3
            RUR1 = 4
        end
    elseif (RU1 > CH1) && (RU1>GB1)
        RUR1 = 2
        if (CH1>GB1)
            CHR1 = 3
            GBR1 = 4
        else
            GBR1 = 3
            CHR1 = 4
        end
    elseif (GB1 > CH1) && (GB1 > RU1)
        GBR1 = 2
        if (RU1>CH1)
            RUR1 = 3
            CHR1 = 4
        else
            CHR1 = 3
            RUR1 = 4
        end
    end
elseif (CH1 > US1) && (CH1 > RU1)  && (CH1>GB1)
    CHR1 = 1
    if (US1 > RU1) && (US1 > GB1)
        USR = 2 
        if (RU1 > GB1) 
            RUR1 = 3
            GBR1 = 4
        else           %GB greater than RU but less than US and CH 
            GBR1 = 3
            RUR1 = 4
        end
    elseif (RU1 > US1) && (RU1>GB1)
        RUR1 = 2
        if (US1>GB1)
            USR1 = 3
            GBR1 = 4
        else
            GBR1 = 3
            USR1 = 4
        end
    elseif (GB1 > US1) && (GB1 > RU1)
        GBR1 = 2
        if (RU1>US1)
            USR1 = 3
            CHR1 = 4
        else
            USR1 = 3
            RUR1 = 4
        end
    end
end
%{
If you got this far well done! You'll have noticed I decided to not include the last two elseifs for RU and GB
ranking first because it looked scary enough. Again see if you can follow the logic of nested if statements.
%}    

USR1 =  1
CHR1 =  2
RUR1 =  3
GBR1 =  4


Obviously this code is terrible and you would __never be asked to write something like this in the exam__, however if you look through it and follow the logic and can understand what's going on then you can understand any if statement that is thrown in your direction and more importantly __create__ any if statement that you require.

Let's do it a better way:

In [43]:
R
[x I] = sort(R,'descend')

R =

   223   259   334
   102   121   153
   121   145   182
   218   266   336

x =

   223   266   336
   218   259   334
   121   145   182
   102   121   153

I =

   1   4   4
   4   1   1
   3   3   3
   2   2   2



x provides the sorted columns of R. 'I' provides the indices. So $x\left(1,1\right)$ corresponds to $R(1,1)$, $x(2,1)$ corresponds to $R(4,1)$, $x(3,1)$ corrsponds to $R(3,1)$ and $x(4,1)$ corresponds to $R(2,1)$ and likewise for the second and third column.

We can now state where each team ranked in each system:

In [49]:
[x I] = sort(R,'descend');
for i = 1:3
    fprintf('System %.0f \n',i)
    % This for loop will iterate each point system (columns of R)
    for j = 1:4
        % This for loop looks through the row of the specific column and ranks teams
        if I(j,i) == 1
            fprintf('US position is %.0f \n',j)
        elseif I(j,i) == 2
            fprintf('GB position is %.0f \n',j)
        elseif I(j,i) == 3
            fprintf('CH position is %.0f \n',j)
        else
            fprintf('RU position is %.0f \n',j)
        end
    end
    fprintf('\n') % An extra enter to split the rank of each system
end

System 1 
US position is 1 
RU position is 2 
CH position is 3 
GB position is 4 

System 2 
RU position is 1 
US position is 2 
CH position is 3 
GB position is 4 

System 3 
RU position is 1 
US position is 2 
CH position is 3 
GB position is 4 



Much nicer! Noticed it depended on what order I entered the medal table data right at the top with the array M.